### ISSUE TITLE
Support for object Detection pipeline and Augmentations

### ISSUE THREAD

**Issue by harsh2ai** [kornia/kornia#1508]
## 🚀 Feature
Support for object detection models and augmentations within Kornia 

## Motivation

I have been using Kornia for a while now and loved its perfomance so I decided to use it for creating my own ocr, but I ran into a couple of issues like the support for augmentations which albumentations does provide but they still have a lot of issues at hand and isn't friendly so I believe that this feature would of great help to community 

## Alternatives
Some alternatives presently are albumentations but that is still not easy to understand for begginers 

## Additional context
A complete end2end object detection pipleine along with custom augmentations would be a great feature
_____________________________________________________________________

**Comment by shijianjian** [comment]
Can you elaborate what are the missing pieces in your experiment? Also, we are working on supporting the detection now. https://github.com/kornia/kornia/pull/1497

/cc @twsl

**Comment by twsl** [comment]
I propose we create an end to end example using mask rcnn as it includes both, bounding boxes and segmentation masks, as input. Preferably with pytorch lightning to reduce boilerplate code.

**Comment by shijianjian** [comment]
@twsl Great. If possible, may you share the demo after #1497 get merged?

**Comment by twsl** [comment]
@shijianjian we might have to add some additional utilities to simplify the workflow, but i can create a draft after xmas.

**Comment by edgarriba** [comment]
@harsh2ai what ocr have you tried ? This is an interesting feature to have discussed at some point with @ducha-aiki .
Regarding the detection algorithm, I’d advocate to make some generic that can support different like yolov4 or more recent ones like detr, our current face detection, etc. Same for training library, we recently started [`kornia.x`](https://kornia.readthedocs.io/en/latest/x.html#kornia.x.ObjectDetectionTrainer) to make easier the training of the networks we provide (not as a generic library). We base in barebones pytorch and rely on Accelerators (which eventually can be switched to torch.x project). We have been trying to make our framework complaint to other libraries and possibly and possibly add examples using ignite, catalyst, fastai, and others, maybe also gradsflow @aniketmaurya. Check also that we are iterating on an a new api for I/O now based on opencv but later supporting our own thing.

**Comment by harsh2ai** [comment]
@edgarriba you can say its more of a generic OCR being built on top of pytorch with Kornia as the middleware for all operations, since that gives more better perfomance in comparison to cv2, I have different models for different purposes effectively which when stacked up bring up an effective OCR model, this is a project which is being pushed forward by me and my friend

For the object detection pipeline I am intending to add the feature where a person can import any backbone model they like from pytorch or timm library then they can call different set of methods to make changes to the model as they please and instantiate a model and use ```kornia.x``` for the training purposes although a significant amount of time is required for rewriting something like albumentations in kornia, the reason why I am suggesting to rewrite albumentations in kornia is Albumentations still has good level of nasty bugs and its better to not depend on them, just for augmentation purposes.


For the OCR part I intend to make a  somewhat generalized OCR solutions like tesseract do exist but it still dosen't supports a lot of features, modern day OCR can do a lot of things like detecting tables and text within them detecting and recognizing text in the wild and so on, the end user can download whatever model they need and achieve the required task, with a option to customize the model as per their needs wherein the user has the liberty to use his own data on our models(since obviusoly we cannot cover all types of solutions) and all they needs to do is  add the input images for training and a list of bounding boxes of region of interest of those train  images rest the text extraction model remains the same thus saving a lot of time

**Comment by harsh2ai** [comment]
> Can you elaborate what are the missing pieces in your experiment? Also, we are working on supporting the detection now. #1497
> 
> /cc @twsl

I can give you one such example let's say we're resizing the images while doing transformations there should be a functionality to resize the bbox accordingly given the dim of image and it should support all formats like Pascal,coco

**Comment by twsl** [comment]
> 
> I can give you one such example let's say we're resizing the images while doing transformations there should be a functionality to resize the bbox accordingly given the dim of image and it should support all formats like Pascal,coco

@harsh2ai that is already possible with [containers](https://kornia.readthedocs.io/en/latest/augmentation.container.html), you just need to supply the correct datakeys. kornia currently supports xyxy and xywh bbox format right now.

**Comment by harsh2ai** [comment]
> > I can give you one such example let's say we're resizing the images while doing transformations there should be a functionality to resize the bbox accordingly given the dim of image and it should support all formats like Pascal,coco
> 
> @harsh2ai that is already possible with [containers](https://kornia.readthedocs.io/en/latest/augmentation.container.html), you just need to supply the correct datakeys. kornia currently supports xyxy (pascal) and xywh (coco) bbox format.

@twsl could you share any link where it's implemented or maybe a tutorial/example regarding the same

**Comment by edgarriba** [comment]
@harsh2ai [`kornia.augmentations`](https://kornia.readthedocs.io/en/latest/augmentation.html) it's already an intend of improvement over existing libraries. Albumentations it's based in opencv -- kornia is opencv in pytorch, so kornia.augmentation should be the seamless library.

Currently, there's an effort to re-organize and refactor that module by @twsl and @shijianjian. If there's any specific feature requests let's just open a ticket and put it in the backlog. Besides, I started this recent project [`kornia_io`](https://github.com/kornia/kornia_io/pull/1) to support camera and image APIs, please check it out and add some comments as well.

Regarding the OCR project -- we recently had a discussion with @anguelos about a potential `kornia.text` module to be the new generation of opencv.text. @anguelos was one of the responsible to refactor and maintain that module few years back in opencv. Possibly it's time to refresh ideas, implement in terms of pytorch and use the knowledge in terms of existing API and @harsh2ai active project. I'd suggest to open a new ticket thread to discuss about that specific top and roadmap.

**Comment by edgarriba** [comment]
@harsh2ai @anguelos for `kornia-text` worth to check on https://github.com/open-mmlab/mmocr to overlap less. I would suggest to find a lightweight network to target embedded devices.

**Comment by harsh2ai** [comment]
@edgarriba I would look up items for this issue of finding lightweight models , also regarding youre previous reply I didn't get what you meant by opening a new ticket thread could you elaborate on what you meant on that part

**Comment by twsl** [comment]
I think the most important part for a clean object detection pipeline are a support for list of tensors as input to support different input sizes and properly respecting the bounding box input format. see https://github.com/kornia/kornia/issues/1522

**Comment by edgarriba** [comment]
@gau-nernst for awareness from previous requests
